File: src/storage.ts
import { readdirSync, statSync, unlinkSync } from "node:fs";
import { join } from "node:path";
import type { Config } from "./types.js";

interface BackupFile {
  name: string;
  path: string;
  mtime: Date;
  sizeBytes: number;
}

function getAgeInDays(date: Date): number {
  const now = new Date();
  const diffMs = now.getTime() - date.getTime();
  return diffMs / (1000 * 60 * 60 * 24);
}

function getStartOfWeek(date: Date): Date {
  const d = new Date(date);
  d.setHours(0, 0, 0, 0);
  d.setDate(d.getDate() - d.getDay());
  return d;
}

export function rotateBackups(
  config: Config,
  serviceName: string
): { kept: number; deleted: number } {
  const serviceBackupDir = join(config.backupPath, serviceName);

  let kept = 0;
  let deleted = 0;

  try {
    const files = readdirSync(serviceBackupDir);
    const backups: BackupFile[] = [];

    for (const file of files) {
      const filePath = join(serviceBackupDir, file);
      const stat = statSync(filePath);
      if (!stat.isFile()) continue;

      backups.push({
        name: file,
        path: filePath,
        mtime: stat.mtime,
        sizeBytes: stat.size,
      });
    }

    const weeklyKept = new Set<string>();

    for (const backup of backups) {
      const ageDays = getAgeInDays(backup.mtime);

      if (ageDays <= config.retentionDays) {
        kept++;
        continue;
      }

      if (ageDays <= config.retentionWeeks * 7) {
        const weekStart = getStartOfWeek(backup.mtime).toISOString();
        const weekKey = `${backup.name.split("_")[0]}_${weekStart}`;

        if (!weeklyKept.has(weekKey)) {
          weeklyKept.add(weekKey);
          kept++;
          continue;
        }
      }

      try {
        unlinkSync(backup.path);
        deleted++;
      } catch {
        kept++;
      }
    }
  } catch {
    // Directory doesn't exist yet, nothing to rotate
  }

  return { kept, deleted };
}

export function rotateAllBackups(
  config: Config,
  serviceNames: string[]
): { service: string; kept: number; deleted: number }[] {
  return serviceNames.map((name) => ({
    service: name,
    ...rotateBackups(config, name),
  }));
}

File: storage.go
package main

import (
	"os"
	"path/filepath"
	"time"
)

type backupFile struct {
	name    string
	path    string
	mtime   time.Time
	size    int64
}

func rotateBackups(config *Config, serviceName string) *RotationResult {
	serviceDir := filepath.Join(config.BackupPath, serviceName)

	var kept, deleted int

	entries, err := os.ReadDir(serviceDir)
	if err != nil {
		return &RotationResult{Service: serviceName}
	}

	var backups []backupFile
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		info, err := entry.Info()
		if err != nil {
			continue
		}
		backups = append(backups, backupFile{
			name:  entry.Name(),
			path:  filepath.Join(serviceDir, entry.Name()),
			mtime: info.ModTime(),
			size:  info.Size(),
		})
	}

	weeklyKept := make(map[string]bool)

	for _, b := range backups {
		ageDays := time.Since(b.mtime).Hours() / 24

		if ageDays <= float64(config.RetentionDays) {
			kept++
			continue
		}

		if ageDays <= float64(config.RetentionWeeks*7) {
			weekStart := getStartOfWeek(b.mtime)
			prefix := "db"
			if len(b.name) > 4 && b.name[:4] == "file" {
				prefix = "files"
			}
			weekKey := prefix + "_" + weekStart.Format("2006-01-02")

			if !weeklyKept[weekKey] {
				weeklyKept[weekKey] = true
				kept++
				continue
			}
		}

		if err := os.Remove(b.path); err != nil {
			kept++
		} else {
			deleted++
		}
	}

	return &RotationResult{
		Service: serviceName,
		Kept:    kept,
		Deleted: deleted,
	}
}

func rotateAllBackups(config *Config, serviceNames []string) []*RotationResult {
	results := make([]*RotationResult, len(serviceNames))
	for i, name := range serviceNames {
		results[i] = rotateBackups(config, name)
	}
	return results
}

func getStartOfWeek(t time.Time) time.Time {
	d := t
	for d.Weekday() != time.Sunday {
		d = d.AddDate(0, 0, -1)
	}
	return time.Date(d.Year(), d.Month(), d.Day(), 0, 0, 0, 0, d.Location())
}
